[Accordion] Replace IconButton wrapper with div

croraf opened the pull request:
### Breaking changes

- [Accordion] Remove `IconButtonProps` prop from AccordionSummary.
  The component renders a `<div>` element instead of an IconButton.
  The prop is no longer necessary.

---

<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [X] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui-org/material-ui/blob/next/CONTRIBUTING.md#sending-a-pull-request).

As per https://github.com/mui-org/material-ui/issues/22804#issuecomment-701046569

---------------
@oliviertassinari 

> However for this demo: https://next.material-ui.com/components/accordion/#additional-actions, we likely want it.

I don't see that this demo is relevant to this change. This demo talks about interactive elements put in the children of the `AccordionSummary` not in the `iconElement`.

---------------------
I think the wrapping element around `iconElement` should still be present, but it's only function will be to fix the `iconElement` to the right and to rotate it around "z-axis" (top/bot), so I removed the wrapping element props completely. I think this is OK, but would like to hear if these wrapping element props should stay for some reason?

--------------------
If we agree to remove wrapping element's props (as per previous paragraph) I should complete the removal by removing the remaining stuff from docs/pages/api-docs/accordion-summary.md and packages/material-ui/src/AccordionSummary/AccordionSummary.test.js

oliviertassinari commented:
> I don't see that this demo is relevant to this change.

In the default Accordion use case, the whole summary is interactive. In this demo, only a limited surface, with the expand icon at its center, is interactive.

croraf commented:
@oliviertassinari 
> In the default Accordion use case, the whole summary is interactive. 

If by the "default Accordion use case" you mean https://next.material-ui.com/components/accordion/#basic-accordion , than the whole summary is interactive, but iconElement has no special interaction behavior, it just passes the interaction to the root AccordionSummary. In this PR this is accounted better than what we currently have.

> In this demo, only a limited surface, with the expand icon at its center, is interactive.

Not sure what you mean by this. The expand icon is on the far right. Or do you mean "center" in the vertical direction?
![image](https://user-images.githubusercontent.com/19698636/94679442-6566ba00-0320-11eb-9223-617b947b448c.png)

mbrookes commented:
@croraf Regarding the additional actions demo - what we loose is the hover indicator on the expand icon. Since most of the area has the action of selecting the checkbox, and the pointer doesn't change, it's less obvious where to click to expand or collapse the panel (particularly with the third panel).

croraf commented:
@mbrookes What do you mean by the hover indicator?

mbrookes commented:
Oh, no you're right. It's disabled anyway. I guess we're good in that case.

croraf commented:
What do you guys think about removing the customization prop for the icon element's wrapper (now it is a div).

>  I think the wrapping element around iconElement should still be present, but it's only function will be to fix the iconElement to the right and to rotate it around "z-axis" (top/bot), so I removed the wrapping element props completely. I think this is OK, but would like to hear if these wrapping element props should stay for some reason?

oliviertassinari commented:
> Oh, no you're right. It's disabled anyway. 

How about we enable the hover for this case?

https://github.com/mui-org/material-ui/blob/345d034c0ab91f3f5cbb0c0e7c75bf0073231eba/packages/material-ui/src/AccordionSummary/AccordionSummary.js#L55-L60

Basically, we could remove the IconButton from `AccordionSummary` but have the demo use it, without any customization.

croraf commented:
I wouldn't put any on-hover in the component. Although it looks nice I don't think this is the intention of the AccordionSummary. If the AccordionSummary is not clickable but "iconElement" is clickable then that would make sense.

The user can put the IconButton with the Icon inside the "iconElement" prop to get this behavior. Did you mean to do that in the demo example?

mbrookes commented:
@croraf Please could you fix the icon color? Not sure it was correct before, but it's further from the spec now.

croraf commented:
@mbrookes Is that the difference shown in argos CI utility?

mbrookes commented:
Yes, although there's also some alignment shift that I didn't dig into. It might be expected after removing the IconButton wrapper

croraf commented:
I checked again https://material.io/archive/guidelines/components/expansion-panels.html and it says 

> Expand icon
> 24dp x 24dp
> 38% black

But below in https://material.io/archive/guidelines/components/expansion-panels.html#expansion-panels-behavior it says:

> Expand icon on hover and pressed
> 24dp x 24dp
> 54% black

croraf commented:
Also how should it behave when disabled is not clear.

I did find an angular implementation: https://material.angular.io/components/expansion/examples

croraf commented:
As per the link on material design "specs" several things are suspicious.

Expand icon (_current_ means before my PR):
1. Our current expand icon color is wrong in idle state. It is 54% black while the spec says 38%.
2. The spec is confusing about focus/hover/click behavior. I pretty sure that all these 3 events happen **only** on the entire "button". But if you see the image below, when describing focus state it doesn't say the behavior of expand icon in that state but in the other two states
![image](https://user-images.githubusercontent.com/19698636/95014779-775e9a80-0649-11eb-8d3e-3b5a919cda28.png)
3. Our current colors do not change when in "focus/hover/click" state, but remain the same as in idle state.

Button:
1. When focused it does not conform to the spec which says Grey 200, which is #EEE , we have black with 12% opacity, which is not the same. Angular impl has black with 4% opacity which is also not the same, as in spec.

(Primary and secondary content I didn't have time to verify.)

Please give suggestion how to proceed.

croraf commented:
I adjusted bullet 1 above and bullet 4 above. Also I adjusted the position of expandIcon to be in the center.

Bullets 2 and 3 remain, but I would like to open a separate issue for them if needed, after these changes are merged.
Also the positions have to be double-checked in a separate issue, as I noticed the current positions are not according to the spec.

oliviertassinari reviewed (approved):
We should be good to go.

Regarding the Material Design spec. The component is no longer included in the guidelines, what we have is outdated. The only place I could find an accordion in Google's product is the search:

<img width="681" alt="Capture d’écran 2020-10-05 à 00 46 33" src="https://user-images.githubusercontent.com/3165635/95028912-50818200-06a4-11eb-8c05-302562fd7a71.png">

Does somebody know another place?

Regarding the additional actions demo. How about we drop it? We nest a button with a checkbox. It's invalid. It would also allow closing #22276.

croraf commented:
@oliviertassinari With your changes you overwrote some things that I explicitly set. The focus color `theme.palette.action.focus` and expandIcon color `theme.palette.action.active` are not correct now as per spec.

oliviertassinari commented:
@croraf The component is no longer in the Material Design guidelines. We should aim for using the design token that makes the most sense in the context, and that would likely result in the best UI once customized. One way to test that is to turn in the dark mode, but it's only the beginning of an answer. In this case. The design token used is the same one as the IconButton.

croraf commented:
OK. You can merge it then. But then, I guess, in the long run the component should be removed completely from MUI.

oliviertassinari commented:
> the component should be removed completely from MUI

Implementing Material Design isn't our primary objective (it was in the beginning). We aim to be a general purpose set of components (at least on the MIT core). It's about maximizing for adoption. Considering the Accordion is a generic component, we will very likely support it indefinitely.
